sns/parity_math.c

```diff
@@ -670,6 +858,168 @@ static void xor_recover(struct m0_parity_math *math,
         }
 }
 
+#ifndef __KERNEL__
+static void isal_recover(struct m0_parity_math *math,
+			 struct m0_buf *data,
+			 struct m0_buf *parity,
+			 struct m0_buf *fails,
+			 enum m0_parity_linsys_algo algo)
+{
+	uint8_t *fail;
+	uint32_t fail_count;
+	uint32_t unit_count = math->pmi_data_count + math->pmi_parity_count;
+	uint32_t block_size = data[0].b_nob;
+	uint8_t *temp_matrix = NULL;
+	uint8_t *invert_matrix = NULL;
+	uint8_t *decode_matrix = NULL;
+	uint8_t **data_in = NULL;
+	uint8_t **data_out = NULL;
+	uint32_t ui, i, j, r;
+	uint8_t *err_list = NULL;
+	uint8_t s;
+	int ret;
+
+	fail = (uint8_t*) fails->b_addr;
+	fail_count = fails_count(fail, unit_count);
+
+#if DEBUG
+	printf("\n fail_count: %d", fail_count);
+#endif
+
+	M0_ASSERT(fail_count > 0);
+	M0_ASSERT(fail_count <= math->pmi_parity_count);
+
+	M0_ALLOC_ARR(err_list, fail_count);
+	M0_ASSERT(err_list != NULL);
+
+	M0_ALLOC_ARR(data_in, (math->pmi_data_count * sizeof(uint8_t *)));
+	M0_ASSERT(data_in != NULL);
+
+	M0_ALLOC_ARR(data_out, (fail_count * sizeof(uint8_t *)));
+	M0_ASSERT(data_out != NULL);
+
+#if DEBUG
+	printf("\n err_list: ");
+#endif
+	for (i = 0, j = 0, r = 0; i < unit_count; i++) {
+		if (fail[i]) {
+			err_list[j] = i;
+
+			if (i < math->pmi_data_count)
+				data_out[j] = math->data_frags[i];
+			else
+				data_out[j] = math->parity_frags[i - math->pmi_data_count];
+
+#if DEBUG
+			printf(" %d", i);
+#endif
+			j++;
+		} else {
+			if (i < math->pmi_data_count)
+				data_in[r] = math->data_frags[i];
+			else
+				data_in[r] = math->parity_frags[i - math->pmi_data_count];
+
+			r++;
+		}
+	}
+
+	for (ui = 1; ui < math->pmi_data_count; ++ui)
+		M0_ASSERT(block_size == data[ui].b_nob);
+
+	for (ui = 0; ui < math->pmi_parity_count; ++ui)
+		M0_ASSERT(block_size == parity[ui].b_nob);
+
+	M0_ALLOC_ARR(temp_matrix, (unit_count * math->pmi_data_count));
+	M0_ASSERT(temp_matrix != NULL);
```

Comment:
All of these error (Out-of-memory) should be handled properly.
No M0_ASSERT() here.

Reply:
Created another ticket to return error codes for these functions.